Win32.xs

```diff
@@ -1682,6 +1693,248 @@ XS(w32_IsDeveloperModeEnabled)
+                           FILE_ATTRIBUTE_NORMAL,
+                           NULL);
+
+        if (!hOut || hOut == INVALID_HANDLE_VALUE)
+            bFileError = TRUE;
+    }
+
+    /* Keep checking for data until there is nothing left. */
+    if (!bFileError && bResults) {
+        do {
+            /* Check for available data. */
+            dwSize = 0;
+            if (!WinHttpQueryDataAvailable(hRequest, &dwSize)) {
+                bAborted = TRUE;
+                break;
+            }
+
+            /* No more available data. */
+            if (!dwSize)
+                break;
+
+            /* Allocate space for the buffer. */
+            New(0, pszOutBuffer, dwSize + 1, char);
+            if (!pszOutBuffer) {
+                bAborted = TRUE;
+                break;
+            }
+
+            /* Read the Data. */
+            ZeroMemory(pszOutBuffer, dwSize+1);
+
+            if (!WinHttpReadData(hRequest,
+                                 (LPVOID)pszOutBuffer,
+                                 dwSize,
+                                 &dwDownloaded)) {
+                bAborted = TRUE;
+                Safefree(pszOutBuffer);
+                break;
+            }
+
+            /* Write what we just read to the output file */
+            if (!WriteFile(hOut,
+                           pszOutBuffer,
+                           dwDownloaded,
+                           &dwBytesWritten,
+                           NULL)) {
+                bAborted = TRUE;
+                bFileError = TRUE;
+                Safefree(pszOutBuffer);
+                break;
+            }
+
+            Safefree(pszOutBuffer);
+
+            /* This condition would only be reached if WinHttpQueryDataAvailable
+             * said there are more data to read but WinHttpReadData didn't get any.
+             */
+            if (!dwDownloaded)
+                break;
+        }
+        while (dwSize > 0);
+    }
+    else {
+        bAborted = TRUE;
+    }
+
+    /* Clean-up may lose this. */
+    if (bAborted)
+        error = GetLastError();
+
+    /* Close any open handles. */
+    if (hOut) CloseHandle(hOut);
+    if (hRequest) WinHttpCloseHandle(hRequest);
+    if (hConnect) WinHttpCloseHandle(hConnect);
+    if (hSession) WinHttpCloseHandle(hSession);
+
+    Safefree(url);
+    Safefree(file);
+    Safefree(hostName);
+    Safefree(urlPath);
+
+    if (bAborted) {
+        char msgbuf[ONE_K_BUFSIZE];
+        DWORD msgFlags = bFileError
+                         ? FORMAT_MESSAGE_FROM_SYSTEM
+                         : FORMAT_MESSAGE_FROM_HMODULE;
+
+        if (FormatMessageA(msgFlags,
+                           GetModuleHandleA("winhttp.dll"),
+                           error,
+                           0,
+                           msgbuf,
+                           sizeof(msgbuf) - 1,
+                           NULL)) {
+            Perl_warn(aTHX_ "Error %lu in Win32::HttpGetFile: %s", error, msgbuf);
```

Comment:
I'm not sure if warning is a good idea. If a user wants to get the error message, they can call Win32::GetLastError. The main problem with non-core warnings is that opting-out from them involves `$SIG{__WARN__}`, which is awkward.

I know some Win32 functions do that too, but IMO it was a mistake.

Reply:
> I'm not sure if warning is a good idea. If a user wants to get the error message, they can call Win32::GetLastError. The main problem with non-core warnings is that opting-out from them involves `$SIG{__WARN__}`, which is awkward.
> 
> I know some Win32 functions do that too, but IMO it was a mistake.

I can see your point.  However, if we don't retrieve the message here the user has no way to get it.  Win32::FormatMessage only gets messages for system errors, not module-specific errors like those from winhttp.  So the programmer would have to call Win32::GetLastError() and then the end user would have to know, for example, that 12005 is an invalid URL.

Reply:
You're right, I forgot about this. Some ideas, sadly none of them are particularly good:

- Store the error code and its source (system, winhttp) in [MY_CXT](https://perldoc.perl.org/perlxs#Safely-Storing-Static-Data-in-XS) and create functions to read and format it.
- Return a special error code (e.g. `"%d:%d"` where the first number is the source and the second is the code) and create a function to format it.
- Return a list with two error codes (system and winhttp) and create a function to format winhttp errors.
- Some kind of a response object?

Reply:
I don't have a good suggestion for the error either, but I can add more questionable ones:

* use `croak` instead of `warn`, so you have to wrap the call in an `eval` block if you want to handle errors yourself
* Return a PVIV, with the error  in the PV slot, and 0 in the IV, so it is still `false` in a boolean context. Feels wrong though.

Reply:
It's easy enough to capture (and throw away) stderr output if people don't want it and I've now done that for the tests that expect to see a warning.  I can't really think of a different way to do this that I like better.

Reply:
>It's easy enough to capture (and throw away) stderr output if people don't want it and I've now done that for the tests that expect to see a warning.

I disagree with the "easy enough" part. E.g. if you don't wrap the code in its own scope, you'll leave STDERR closed and the output of e.g. `die` later in the same scope would be swallowed. I think handling this correctly is somewhat verbose and can easily introduce errors.

>I can't really think of a different way to do this that I like better.

I would like you to comment on my suggestion to use `croak` instead. And I would also like to see some feedback on this issue from @xenu as well!

Reply:
I think the most likely thing someone will do with the croak/eval approach is `warn $@` and continue, which ends up being essentially what we already have with the warning.  Remember the nominal "customer" is a CPAN client that is going to try to download a file via a list of methods, probably IO::Socket::SSL first if available, then maybe spawning curl or wget, and then somewhere in the list, if on Windows, possibly (in the future), this new function.  This new function may not be the last in the list, so if it fails, continuing down the list and trying something else may very well happen. But surely the user would get some message about why this didn't work before the caller moves on to try some other download method.

All that said, I don't do heavy-duty server-side code in Perl and in fact rarely get to do much Perl at all lately, so I may be out of touch with best practices.  If croak/eval is really how this sort of thing is usually done, I don't have a major objection to doing that instead of warning.  Let me know if you'd like me to proceed with that change.

Reply:
I'm not sure either, which is why I'm asking @xenu to chime in as well.

Personally I just really dislike library code writing warnings to STDERR instead of returning proper error information and letting the caller decide how to handle the error condition.

So to me returning 2 values in list context (the success status, and an error message) would be another option, but I feel like using `eval` leads to more compact code:

```perl
my($ok, $err) = Win32::HttpGetFile(...);
unless ($ok) {
    # error handling, can use $err and Win32::GetLastError()
}
```

It would still allow for

```perl
unless (Win32::HttpGetFile(...)) {
    # error handling, but all we have is Win32::GetLastError()
}
```

With this you can still get the current behavior:

```perl
my($ok, $err) = Win32::HttpGetFile(...);
warn $err unless $ok;
```

The `eval` version would look like this instead:

```perl
unless (eval { Win32::HttpGetFile(...) }) {
    # error handling, we have $@ in place of $err, but also Win32::GetLastError()
}
```

And the current functionality would be:

```perl
eval { Win32::HttpGetFile(...) };
warn $@ if $@;
```

Now that I wrote all this down, I think I may prefer the "return 2 values in list context" version to the code requiring `eval`.

But let's wait for @xenu to give heir feedback, and/or maybe solicit some feedback from p5p?

Reply:
Unfortunately, it seems that providing an HTTP client API is a difficult problem, so we can't get away with returning just a boolean. None of the popular HTTP clients do this (HTTP::Tiny, LWP::UserAgent, Mojo::UserAgent). LWP::Simple **does** have a simple API, but because of complete lack of error handling many people consider it useless.

Also, I'm afraid we forgot about something very important: HTTP response codes, especially 4xx and 5xx.

So this is my latest proposal. It's a bit verbose but I don't think it's that much worse than HTTP::Tiny.

```perl
# always a list, not context sensitive
# the last parameter is *optional* and means "ignore SSL errors", default is "false"
my($error_code, $http_code) = Win32::HttpGetFile("https://www.example.com", "file.txt", 1);
if ($error_code == 0) {
    if ($http_code >= 200 && $http_code < 300) {
        # success
    }
    else {
        unlink("file.txt");
	die "HTTP response code: $http_code";
    }
}
# Windows error codes are unsigned, which means we can use negative values for our
# own purposes. Let's return winhttp error codes as negative values.
elsif ($error_code > 0) {
    die "System error: " . Win32::FormatMessage($error_code);
}
else {
    # new function
    die "WinHTTP error: " . Win32::FormatWinHTTPMessage(abs $error_code);
}
```

Reply:
>I think the most likely thing someone will do with the croak/eval approach is warn $@ and continue, which ends up being essentially what we already have with the warning.

They may want to log it somewhere, prefix the message with something or even recover from the error by e.g. using a different server.

Reply:
I disagree with the added complexity. This is not a general-purpose HTTP client library, it is a single function with a singular purpose: fetch a resource via http or https and store it in the filesystem. It either succeeds, or it doesn't. If it doesn't, I would expect LastError to have some reasonable reason for this.

Where does the additional complexity provide any benefit?

PS: I especially object to adding an extra `Win32::FormatWinHTTPMessage` API to the core module; that feels like a bad hack.

Reply:
> `# the last parameter is *optional* and means "ignore SSL errors", default is "false"`

I agree that this would be useful, assuming "ignore SSL errors" means "don't try to validate the certificate(s)". Any other errors should still be fatal.

Reply:
> it is a single function with a singular purpose: fetch a resource via http or https and store it in the filesystem

Put another way, it is a replacement for

```shell
curl -o $FILENAME $URL
```

If `$?` is `0`, it was a success, otherwise it was a failure, and the stdout/stderr contains more information when it failed. `Win32::HttpGetFile` should not be more complex than this.

Reply:
> If it doesn't, I would expect LastError to have some reasonable reason for this.

>PS: I especially object to adding an extra Win32::FormatWinHTTPMessage API to the core module; that feels like a bad hack.

The problem is that our new function can either fail at something like CreateFile() (which uses windows error codes) or inside WinHTTP. You can't use `Win32::FormatMessage` for the latter.

>curl -o $FILENAME $URL

What about http status codes? by default curl ignores e.g. 404 and happily downloads the error message to disk while wget aborts.

Reply:
>I agree that this would be useful, assuming "ignore SSL errors" means "don't try to validate the certificate(s)". Any other errors should still be fatal.

Agreed.

Reply:
>The problem is that our new function can either fail at something like CreateFile() (which uses windows error codes) or inside WinHTTP. You can't use Win32::FormatMessage for the latter.

Both the `croak` version and the `return 2 values in list context` version already make the error text available, either as `$@` or as a return value. Asking the user to call `FormatMessage` is exposing implementation details that don't matter to the user.

>What about http status codes? by default curl ignores e.g. 404 and happily downloads the error message to disk while wget aborts.

I have not checked the implementation; I would expect that any status code outside 200-299 would result in a failure and set a proper error code. This should be added to the tests!  I would also expect that an incomplete download would result in a failure. Our function should be equivalent to

```
curl --silent --fail --show-error -o $FILE $URL
```

with the error information available via the error message and not on stderr.

Reply:
I agree might proposal was a bit too complex. It's unlikely that users will be able to anything useful with windows/winhttp error codes, so when they happen, croaking is probably good enough (BTW, Mojo::UserAgent does exactly that).

The main problem are HTTP codes. We can't just pretend they don't exist. However, it's unclear whose responsibility is to handle them. 

I see two options:

- We could copy wget and treat non-2xx return codes as fatal errors and croak with whatever WINHTTP_QUERY_STATUS_TEXT returns.

- Leave it to the user: always download the response and return the HTTP code. Or maybe a list/hash "is_success", "error_code".

If maximum simplicity is our goal, the first option is probably the best.